File: 2024/kotlin/src/day03/Main.kt
package day03

import java.io.File

fun solvePartOne(input: String): Int {
  val pattern ="""mul\((\d{1,3}),(\d{1,3})\)""".toRegex()
  return pattern
    .findAll(input)
    .map { matchResult ->
      // Extract captured groups and multiply them
      val (first, second) = matchResult.destructured
      first.toInt() * second.toInt()
    }
    .sum()
}

fun solvePartTwo(input: String): Int {
  var instructionEnabled = true
  var solutionTwoSum = 0

  val pattern = Regex("""(do|don't)\(\)|mul\((\d{1,3}),(\d{1,3})\)""")
  pattern.findAll(input).forEach { match ->
    val (instr, first, second) = match.destructured
    when {
      instr == "don't" -> instructionEnabled = false
      instr == "do" -> instructionEnabled = true
      instructionEnabled -> {
        solutionTwoSum += first.toInt() * second.toInt()
      }
    }
  }
  
  return solutionTwoSum
}

fun main() {
  val input = File("src/day03/input.txt").readText()
  println("Part 1: ${solvePartOne(input)}")
  println("Part 2: ${solvePartTwo(input)}")
}

File: 2024/javascript/day03/main.js
import { readFileSync } from 'node:fs';

// Read values from file
const input = readFileSync('./input.txt').toString();

let sum = 0;
const pattern = /mul\((\d+),(\d+)\)/g;
for (const match of input.matchAll(pattern)) {
  sum += Number(match[1]) * Number(match[2])
}
console.log(sum);

let sum2 = 0;
let enabled = true;
const pattern2 = /(do|don\'t)\(\)|mul\((\d+),(\d+)\)/g;
for (const match of input.matchAll(pattern2)) {
  if (match[1] === "don't") {
    enabled = false;
    continue;
  } 
  if (match[1] === "do") {
    enabled = true;
    continue;
  }
  if (!enabled) {
    continue;
  }
  // console.log(match[1], match[2])
  sum2 += Number(match[2]) * Number(match[3])
}
console.log(sum2);
